modules/resourceid/fsw/src/cfe_resourceid_api.c

```diff
@@ -0,0 +1,133 @@
+**
+**  Unless required by applicable law or agreed to in writing, software
+**  distributed under the License is distributed on an "AS IS" BASIS,
+**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**  See the License for the specific language governing permissions and
+**  limitations under the License.
+*/
+
+/*
+**  File:
+**  cfe_resource_api.c
+**
+**  Purpose:
+**      Function definitions related to CFE resource management
+**
+**  References:
+**     Flight Software Branch C Coding Standard Version 1.0a
+**     cFE Flight Software Application Developers Guide
+*/
+
+/*
+** Includes
+*/
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "cfe.h"
+#include "cfe_resourceid.h"
+#include "cfe_resourceid_basevalue.h"
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_GetBase
+ *
+ * For complete API information, see prototype in header
+ */
+uint32 CFE_ResourceId_GetBase(CFE_ResourceId_t ResourceId)
+{
+    return (CFE_ResourceId_ToInteger(ResourceId) & ~((uint32)CFE_RESOURCEID_MAX));
+}
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_GetSerial
+ *
+ * For complete API information, see prototype in header
+ */
+uint32 CFE_ResourceId_GetSerial(CFE_ResourceId_t ResourceId)
+{
+    return (CFE_ResourceId_ToInteger(ResourceId) & ((uint32)CFE_RESOURCEID_MAX));
+}
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_ToIndex
+ *
+ * For complete API information, see prototype in header
+ */
+int32 CFE_ResourceId_ToIndex(CFE_ResourceId_t Id, uint32 BaseValue, uint32 TableSize, uint32 *Idx)
+{
+    uint32 Serial;
+
+    if (Idx == NULL)
+    {
+        return CFE_ES_BAD_ARGUMENT;
+    }
+
+    Serial = CFE_ResourceId_ToInteger(Id) - BaseValue;
+
+    if (Serial > CFE_RESOURCEID_MAX || TableSize == 0)
+    {
+        return CFE_ES_ERR_RESOURCEID_NOT_VALID;
+    }
+
+    *Idx = Serial % TableSize;
+    return CFE_SUCCESS;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ResourceId_FindNext
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ResourceId_t CFE_ResourceId_FindNext(CFE_ResourceId_t StartId, uint32 TableSize,
+                                         bool (*CheckFunc)(CFE_ResourceId_t))
+{
+    uint32           Serial;
+    uint32           Count;
+    uint32           ResourceType;
+    CFE_ResourceId_t CheckId;
+    bool             IsTaken;
+
+    ResourceType = CFE_ResourceId_GetBase(StartId);
+    Serial       = CFE_ResourceId_GetSerial(StartId);
+
+    Count   = TableSize;
+    IsTaken = true;
+
+    do
+    {
+        if (Count == 0)
+        {
+            CheckId = CFE_RESOURCEID_UNDEFINED;
+            break;
+        }
+
+        --Count;
+        ++Serial;
+        if (Serial >= CFE_RESOURCEID_MAX)
+        {
+            Serial %= TableSize;
```

Comment:
I'm unsure that we need TableSize in this function, though I'm not clear on its use. I'm assuming TableSize should be the size of the resource ID table for each resource ID type, but I thought each table section should contain up to CFE_RESOURCE_ID_MAX IDs since their bases are all allocated on those boundaries by CFE_RESOURCEID_MAKE_BASE. Line 123 checks to see if we've gone over the table boundary by checking if the serial for the current ID is larger than CFE_RESOURCEID_MAX.  It then causes Serial to wrap around using the modulo operator with the supplied TableSize value. If TableSize is CFE_RESOURCEID_MAX, then Serial will be 0 which should be the start of the table. If we specified TableSize as a value smaller than CFE_RESOURCE_ID_MAX, then modulo will reset the serial number to some other place in the table entirely between 0 and TableSize after it rolls over. For instance, if we specified TableSize as 200 and we're trying to find the next ID and it rolls over the end of the table at 65535, then it will start checking again starting at serial number 35. 

We could maybe instead get away with just setting Serial to 0 on line 125 and setting Count to CFE_RESOURCEID_MAX.

Reply:
Similar questions came up during OSAL review.  

Resource ID serial numbers and table positions are not the same thing.  Serial numbers change/increase each time a new object is allocated, up to `CFE_RESOURCEID_MAX`, whereas the `TableSize` is the actual number of objects that can coexist at the same time.  

The intent is that when an object is deleted and then a new object is created, it may re-use the same memory slot/table position that the old object had, but the new object does __not__ get the same ID as the old object had.

Since the number space is not infinite serial numbers will eventually be recycled but it takes thousands of allocations and deletions before that happens.

Reply:
Note that this is structured such that table position is always mathematically calculated/fixed based on serial number.   
Given a serial number (i.e. extracted from a passed in ID), taking modulus with the size of the table results in the expected/correct table position for that object.  

The reverse option cannot be mathematically computed, one has to go to the entry in the object table and get the resource ID out of it.